File: C++/代码随想录/Linked_List/160. Intersection of Two Linked Lists.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        if not headA or not headB:
            return None

        pointerA, pointerB = headA, headB
        
        while pointerA != pointerB:
            pointerA = pointerA.next if pointerA else headB
            pointerB = pointerB.next if pointerB else headA

        return pointerA

sol = Solution()

# 测试用例
# 创建两个链表：A: 1->2->3->4->5 和 B: 6->7->8->3->4->5 (相交于节点3)
shared = ListNode(3, ListNode(4, ListNode(5)))
listA = ListNode(1, ListNode(2, shared))
listB = ListNode(6, ListNode(7, ListNode(8, shared)))

print(sol.getIntersectionNode(listA, listB).val)

File: C++/代码随想录/Linked_List/160. Intersection of Two Linked Lists.cpp
// 160. Intersection of Two Linked Lists
// https://leetcode.com/problems/intersection-of-two-linked-lists/description/

#include "LinkedListUtils.hpp"

ListNode* getIntersectionNode(ListNode* headA, ListNode* headB) {
    if (!headA || !headB) return nullptr;
    
    ListNode* pA = headA;
    ListNode* pB = headB;
    
    while (pA != pB) {
        pA = pA ? pA->next : headB;
        pB = pB ? pB->next : headA;
    }
    
    return pA;
}

int main() {
    ListNode* headA = new ListNode(4);
    ListNode* node1 = new ListNode(1);
    ListNode* node8 = new ListNode(8);
    ListNode* node4 = new ListNode(4);
    ListNode* node5 = new ListNode(5);
    headA->next = node1;
    node1->next = node8;
    node8->next = node4;
    node4->next = node5;
    
    ListNode* headB = new ListNode(5);
    ListNode* node0 = new ListNode(0);
    ListNode* node1b = new ListNode(1);
    headB->next = node0;
    node0->next = node1b;
    node1b->next = node8; // Intersection point
    
    ListNode* intersection = getIntersectionNode(headA, headB);
    
    if (intersection != nullptr) {
        std::cout << "Intersection point: " << intersection->val << std::endl;
    } else {
        std::cout << "No intersection point found." << std::endl;
    }
    
    return 0;
}
